File: src/model_story/story_engine/phases.py
from __future__ import annotations

from model_story.enums import Phase


def compute_phase(
    progress: float,
    primary_value: float,
    baseline: float,
    target: float,
) -> Phase:
    """Hybrid phase detector using both progress and relative metric improvement.

    Args:
        progress:      0.0–1.0 fraction of total epochs completed.
        primary_value: Current primary metric value (accuracy, mAP, etc.).
        baseline:      Value at epoch 0 (or first recorded epoch).
        target:        Theoretical maximum (1.0 for accuracy/mAP, or pre-configured).
    """
    if progress < 0.10:
        return Phase.AWAKENING

    span = target - baseline
    relative = (primary_value - baseline) / (span + 1e-9) if span > 0 else progress

    if progress < 0.40 or relative < 0.40:
        return Phase.LEARNING
    if progress < 0.70 or relative < 0.75:
        return Phase.UNDERSTANDING
    if progress < 0.95 or relative < 0.95:
        return Phase.MASTERING
    return Phase.POLISHING


def estimate_progress(
    current_epoch: float | None,
    total_epochs: int | None,
    step: int | None = None,
    total_steps: int | None = None,
) -> float:
    """Best-effort 0–1 progress estimate from whatever info is available."""
    if current_epoch is not None and total_epochs and total_epochs > 0:
        return min(current_epoch / total_epochs, 1.0)
    if step is not None and total_steps and total_steps > 0:
        return min(step / total_steps, 1.0)
    # Fall back to a small non-zero value so AWAKENING fires immediately
    return 0.05

File: model-story-vscode/src/story/phases.ts
/**
 * TypeScript port of src/model_story/story_engine/phases.py
 */

export type Phase =
  | "awakening"
  | "learning"
  | "understanding"
  | "mastering"
  | "polishing";

export const PHASE_EMOJIS: Record<Phase, string> = {
  awakening: "🌱",
  learning: "📚",
  understanding: "💡",
  mastering: "🎯",
  polishing: "✨",
};

/**
 * Hybrid phase detector using both progress and relative metric improvement.
 *
 * @param progress      0.0–1.0 fraction of total epochs completed.
 * @param primaryValue  Current primary metric value.
 * @param baseline      Value at epoch 0 (or first recorded epoch).
 * @param target        Theoretical maximum (1.0 for accuracy/mAP, etc.).
 */
export function computePhase(
  progress: number,
  primaryValue: number,
  baseline: number,
  target: number,
): Phase {
  if (progress < 0.10) return "awakening";

  const span = target - baseline;
  const relative =
    span > 0 ? (primaryValue - baseline) / (span + 1e-9) : progress;

  if (progress < 0.40 || relative < 0.40) return "learning";
  if (progress < 0.70 || relative < 0.75) return "understanding";
  if (progress < 0.95 || relative < 0.95) return "mastering";
  return "polishing";
}

/**
 * Best-effort 0–1 progress estimate from whatever info is available.
 */
export function estimateProgress(
  currentEpoch: number | null,
  totalEpochs: number | null,
  step: number | null = null,
  totalSteps: number | null = null,
): number {
  if (currentEpoch !== null && totalEpochs !== null && totalEpochs > 0) {
    return Math.min(currentEpoch / totalEpochs, 1.0);
  }
  if (step !== null && totalSteps !== null && totalSteps > 0) {
    return Math.min(step / totalSteps, 1.0);
  }
  return 0.05; // non-zero so AWAKENING fires immediately
}
